File: python/y.py
# y(f)(a) = f(y(f))(a)
y = lambda f: (lambda x: lambda a: f(x(x))(a)) (lambda x: lambda a: f(x(x))(a))

partialFactorial = lambda slf: lambda a: a * slf(a - 1) if a else 1

factorial = y(partialFactorial)

print(factorial(9))

# recursion without assigning names to functions
print(
  (lambda f: (lambda x: lambda a: f(x(x))(a)) (lambda x: lambda a: f(x(x))(a)))
  (lambda slf: lambda a: a * slf(a - 1) if a else 1)
  (9)
)

# compact version
print(
  (lambda f: (lambda x: x(x)) (lambda x: lambda a: f(x(x))(a)))
  (lambda slf: lambda a: a * slf(a - 1) if a else 1)
  (9)
)

File: haskell/y.hs
import Unsafe.Coerce
uc = Unsafe.Coerce.unsafeCoerce

-- y f = f (y f)
y :: (t -> t) -> t
y f = (\x -> f (uc x x)) (\x -> f (uc x x))

partialFactorial :: (Int -> Int) -> Int -> Int
partialFactorial _ 0 = 1
partialFactorial self x = x * self (x - 1)

factorial :: Int -> Int
factorial = y partialFactorial

f9 :: Int
f9 = factorial 9

-- recursion without assigning names to functions
f9' :: Int
f9' = (\f -> (\x -> f (uc x x)) (\x -> f (uc x x)))
  (\slf -> \x -> if x == 0 then 1 else x * slf (x - 1)) 9

-- compact version
f9'' :: Int
f9'' = (\f -> (\x -> uc x x) (\x -> f (uc x x)))
  (\slf -> \x -> if x == 0 then 1 else x * slf (x - 1)) 9

main = print $ [f9, f9', f9'']
